### ISSUE TITLE
Review popular (web/JS) open source projects for common styles

### ISSUE THREAD

**Issue by jzaefferer** [jzaefferer/commitplease#62]
Project needs to use npm for installing dependencies, doesn't really matter what it does otherwise.

JSHint has something vaguely similar to Angular: https://github.com/jshint/jshint/commits/master

ESLint seems pretty similar to jQuery: https://github.com/eslint/eslint/commits/master

etc. pp.

**Comment by all3fox** [comment]
I have ordered the following list by how useful, imo, commitplease might be for the project. First four look rapidly developing with quite a few people working together, so I think we need to start with those.

 1. [Ember](https://github.com/emberjs/ember.js)
    1. [CONTRIBUTING.md](https://github.com/emberjs/ember.js/blob/master/CONTRIBUTING.md#pull-requests) has special prefixes for commit messages 
    1. Using this in other repos, too, like [ember-data](https://github.com/emberjs/data)
 1. [ReactJS](https://github.com/facebook/react)
    1. [Contributing](https://github.com/facebook/react/blob/master/CONTRIBUTING.md) does not mention commit style
    1. Mostly link to issue/pull request in first line
 1. [airbnb/javascript](https://github.com/airbnb/javascript)
    1. No contributing, no commit style discussion
    1. Lead contributor alone uses sort of tags in first line
    1. No contributing instructions, no website, no wiki -> just do it
 1. [Chart.js](https://github.com/chartjs/Chart.js)
    1. [Contributing](https://github.com/chartjs/Chart.js/blob/master/CONTRIBUTING.md#pull-requests) asks for clear commit messages
    1. No consistent message style
 1. [Modernizr](https://github.com/Modernizr/Modernizr)
    1. Did not find any instructions for contributing, only using
    1. Sometimes angular-like commit style
    1. Sometimes issue number in first line
 1. [Backbone.js](https://github.com/jashkenas/backbone)
    1. [Contributing](https://github.com/jashkenas/backbone/blob/master/CONTRIBUTING.md) does not mention commit style
    1. No commit style discussion found
    1. Mostly a capitalized verb in imperative starts the message
    1. Sometimes first line includes link to issue/pr
 1. [Jest](https://github.com/facebook/jest)
    1. [Contributing](https://github.com/facebook/jest/blob/master/CONTRIBUTING.md) does not mention commit style
    1. Mostly link to issue/pull request in first line
 1. [impress.js](https://github.com/impress/impress.js)
    1. [Contributing](https://github.com/impress/impress.js/blob/master/.github/CONTRIBUTING.md) only specifies a “release” commit message
    1. No commit style discussion found
    1. No explicit commit message style guide. Capitalize and use a verb in imperative is the de-facto standard
 1. [vue.js](https://github.com/vuejs/vue)
    1. [Contributing](https://github.com/vuejs/vue/blob/dev/CONTRIBUTING.md) does not mention commit style
    1. Sometimes capitalized (or not) verb in imperative
    1. Sometimes issue number in first line, in parens
    1. Sometimes a [build] or [release] tag
 1. [h5bp](https://github.com/h5bp/html5-boilerplate)
    1. [Contributing](https://github.com/h5bp/html5-boilerplate/blob/master/CONTRIBUTING.md) only asks for a “clear title and description”
    1. No commit style discussion found
    1. Mostly a capitalized verb in imperative starts the message
    1. Sometimes “Doc:” or “Docs:” start a doc-change commit
    1. Sometimes [skip ci] tag at the end of first line
 1.  ~~[node](https://github.com/nodejs/node)~~
    1. ~~[Contributing](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-3-commit) describes a commit style close to jQuery~~
 1.  ~~[Meteor](https://github.com/meteor/meteor)~~
    1. ~~[Contibuting](https://github.com/meteor/meteor/blob/devel/Contributing.md#submitting-pull-requests) links to a style guide for code style and commit messages. Only code style is actually described there~~
    1. ~~Lead developer capitalizes commit message and ends first line with a dot. That is the de-facto commit message style.~~

**Comment by jzaefferer** [comment]
Sounds good, let's start with those four. Could you summarize here what you would suggest to each? Like style or other options.

**Comment by jzaefferer** [comment]
Btw. what about jshint and eslint (I mentioned them above)?

**Comment by all3fox** [comment]
I have added both jshint and eslint to the list above. Both these projects actually describe a style for commit messages. In the case of eslint there is even [automatic checking][1] for that style (that has extra corner cases compared to style guide).

So, I do not know what the best approach would be here:
 1. We could try to convince them to switch to one of the styles that we already support. Probably jQuery style suits both jshint and eslint best.
 1. We could implement support for those styles in commitplease and simply let jshint add automatic checking and elsint -- drop some of their in-house scripts. This is probably the way to go?

[1]: https://github.com/eslint/eslint/blob/cb90359d965e4a18284f58992e8ea75580ae21ed/Makefile.js#L887-L948

**Comment by jzaefferer** [comment]
The eslint script runs as part of a build (I guess), which commitplease doesn't do by default. They could use the commitplease validation module though, instead of implementing their own logic. So I think for eslint we should start with a discussion. See if they're interested in adopting commitplease (it provides more immediate feedback while committing!). If yes, if they want to keep their current format or would consider adopting another. If they want to keep theirs we can consider adding support in commitplease. Though I see that more of a last resort.

More or less the same for jshint: Open an issue where you suggest they adopt the jQuery style with commitplease.

**Comment by all3fox** [comment]
`grep -r Makefile.js` from the repo root shows that `Makefile.js` is just a collection of functions that are called manually by running various `npm run some-script`. In particular, I am pretty sure that `checkGitCommit` is run manually and not actually different from a standalone `validate-commit-msg.js` script that we saw in Angular. [Here is my take](https://gist.github.com/all3fox/049f2a6a064d918a07db369455102c99) at the issue text for eslint, please review.

**Comment by jzaefferer** [comment]
That looks great!

Only nitpick: commitplease is used by jQuery Core, jQuery UI, jQuery Mobile, Globalize and QUnit. Probably others, but npm doesn't allow search for dev-depended-on packages. Linking to the jQuery Core PR makes it look like they aren't actually using it, yet.

**Comment by all3fox** [comment]
Thanks! I have addressed the nitpick and [created the issue](eslint/eslint/issues/6706).

**Comment by all3fox** [comment]
Created a PR to emberjs/ember.js#13893

**Comment by all3fox** [comment]
Created an issue to airbnb/javascript#965

**Comment by all3fox** [comment]
Created an issue to Modernizr/Modernizr#2050

**Comment by all3fox** [comment]
Created an issue to chartjs/Chart.js#3147

**Comment by all3fox** [comment]
Created an issue to jashkenas/backbone#4065

**Comment by jzaefferer** [comment]
I think the outreach last year was plenty. Adoption wasn't anywhere near to what we've hoped for, but that's okay.
